File: src/automark.js
// Auto-marking.
//
// Marking without a tap is only honest if the app is running *because you are in the class*,
// so the app must not be able to keep marking long after you walked away. Two guards do that:
//
//   1. The daily shutdown (SHUTDOWN_AT) stops the server at the end of the teaching day.
//   2. Auto-marking only ever applies on the calendar day it was armed, so a server that
//      somehow survives midnight cannot mark you into tomorrow's 9:30 class.
//
// AUTO_MARK_HOURS optionally adds a shorter expiry on top; 0 means "for as long as the app
// is open", which is the usual setup alongside a daily shutdown.

const HOUR = 3_600_000;

const sameDay = (a, b) =>
  a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth() && a.getDate() === b.getDate();

export class AutoMarker {
  /**
   * @param enabled   arm immediately on start (AUTO_MARK=1)
   * @param windowMs  expiry for an arming; <= 0 means no expiry
   * @param clock     injectable for tests
   */
  constructor({ enabled = false, windowMs = 0, clock = () => Date.now() } = {}) {
    this.windowMs = windowMs > 0 ? windowMs : 0;
    this.clock = clock;
    this.armedAt = enabled ? clock() : null;
  }

  arm() { this.armedAt = this.clock(); return this.status(); }

  disarm() { this.armedAt = null; return this.status(); }

  get expiresAt() {
    if (this.armedAt === null || !this.windowMs) return null;
    return this.armedAt + this.windowMs;
  }

  /** Armed, still inside any expiry, and still the same day it was armed. */
  isActive() {
    if (this.armedAt === null) return false;
    const now = this.clock();
    if (this.windowMs && now >= this.armedAt + this.windowMs) return false;
    return sameDay(new Date(now), new Date(this.armedAt));
  }

  /** Armed at some point, but no longer active. */
  isExpired() { return this.armedAt !== null && !this.isActive(); }

  /** Why it stopped, for the UI and the log. */
  reason() {
    if (this.armedAt === null) return 'off';
    if (this.isActive()) return 'active';
    if (this.windowMs && this.clock() >= this.armedAt + this.windowMs) return 'window-elapsed';
    return 'new-day';
  }

  msRemaining() {
    if (!this.isActive() || !this.windowMs) return null;
    return this.armedAt + this.windowMs - this.clock();
  }

  status() {
    return {
      active: this.isActive(),
      expired: this.isExpired(),
      armed: this.armedAt !== null,
      reason: this.reason(),
      // null expiresAt means "stays on while the app is open"
      expiresAt: this.expiresAt === null ? null : new Date(this.expiresAt).toISOString(),
      msRemaining: this.msRemaining(),
      windowHours: this.windowMs ? this.windowMs / HOUR : 0,
    };
  }
}

/**
 * Milliseconds until the next occurrence of "HH:MM" local time.
 * Returns null if the spec is blank or malformed.
 */
export function msUntilTimeOfDay(spec, now = new Date()) {
  const m = /^(\d{1,2}):(\d{2})$/.exec(String(spec || '').trim());
  if (!m) return null;
  const hour = Number(m[1]);
  const minute = Number(m[2]);
  if (hour > 23 || minute > 59) return null;
  const target = new Date(now);
  target.setHours(hour, minute, 0, 0);
  if (target <= now) target.setDate(target.getDate() + 1);
  return target - now;
}

File: android/app/src/main/java/org/isdm/companion/domain/AutoMark.kt
package org.isdm.companion.domain

import java.time.Clock
import java.time.Duration
import java.time.Instant
import java.time.ZoneId

private const val HOUR_MS = 3_600_000L
private val TIME_OF_DAY_RE = Regex("^(\\d{1,2}):(\\d{2})$")

/**
 * Pure auto-mark guard.  It deliberately does not persist its arming: a process restart must
 * not silently begin marking a new class day.  The policy is clock-injected for deterministic
 * tests and always compares calendar days in [LMS_ZONE].
 */
class AutoMarkPolicy(
    enabled: Boolean = false,
    windowMs: Long = 0L,
    private val clock: Clock = Clock.systemUTC(),
    private val zone: ZoneId = LMS_ZONE,
) {
    private val windowMs: Long = windowMs.coerceAtLeast(0L)
    private var armedAt: Instant? = if (enabled) clock.instant() else null

    fun arm(): AutoMarkStatus {
        armedAt = clock.instant()
        return status()
    }

    fun disarm(): AutoMarkStatus {
        armedAt = null
        return status()
    }

    val expiresAt: Instant?
        get() = armedAt?.let { armed ->
            if (windowMs == 0L) null else armed.plusMillis(windowMs)
        }

    fun isActive(): Boolean = isActiveAt(clock.instant())

    /** Armed at some point, but no longer active. */
    fun isExpired(): Boolean = armedAt != null && !isActive()

    fun reason(): AutoMarkReason = reasonAt(clock.instant())

    /** Milliseconds remaining, or null when inactive/unbounded. */
    fun remainingMs(): Long? {
        val now = clock.instant()
        return remainingMsAt(now)
    }

    fun status(): AutoMarkStatus {
        val now = clock.instant()
        val armed = armedAt != null
        val active = isActiveAt(now)
        val expires = armedAt?.let { armedAtValue ->
            if (windowMs == 0L) null else armedAtValue.plusMillis(windowMs)
        }
        return AutoMarkStatus(
            active = active,
            expired = armed && !active,
            armed = armed,
            reason = reasonAt(now),
            expiresAt = expires,
            remainingMs = remainingMsAt(now),
            windowHours = windowMs / HOUR_MS.toDouble(),
        )
    }

    private fun isActiveAt(now: Instant): Boolean {
        val armed = armedAt ?: return false
        if (windowMs != 0L && !now.isBefore(armed.plusMillis(windowMs))) return false
        return now.atZone(zone).toLocalDate() == armed.atZone(zone).toLocalDate()
    }

    private fun reasonAt(now: Instant): AutoMarkReason {
        val armed = armedAt ?: return AutoMarkReason.OFF
        if (isActiveAt(now)) return AutoMarkReason.ACTIVE
        if (windowMs != 0L && !now.isBefore(armed.plusMillis(windowMs))) {
            return AutoMarkReason.WINDOW_ELAPSED
        }
        return AutoMarkReason.NEW_DAY
    }

    private fun remainingMsAt(now: Instant): Long? {
        if (!isActiveAt(now) || windowMs == 0L) return null
        return Duration.between(now, armedAt!!.plusMillis(windowMs)).toMillis()
    }
}

/**
 * Milliseconds until the next local occurrence of HH:MM.  Blank or malformed input returns
 * null.  Equality rolls to tomorrow, matching the desktop shutdown helper.
 */
fun msUntilTimeOfDay(
    spec: String?,
    now: Instant = Instant.now(),
    zone: ZoneId = LMS_ZONE,
): Long? {
    val match = TIME_OF_DAY_RE.matchEntire(spec.orEmpty().trim()) ?: return null
    val hour = match.groupValues[1].toIntOrNull() ?: return null
    val minute = match.groupValues[2].toIntOrNull() ?: return null
    if (hour > 23 || minute > 59) return null

    val localNow = now.atZone(zone)
    var target = localNow.toLocalDate().atTime(hour, minute).atZone(zone)
    if (!target.toInstant().isAfter(now)) target = target.plusDays(1)
    return Duration.between(now, target.toInstant()).toMillis()
}
